### PR TITLE
[Validator] Add Isin validator constraint

### LINKED ISSUE #36362: ISIN code validator
There is a BIC constraint and validator, same about IBAN constraint and validator.

What do you think about adding an ISIN constraint and validator ?

Validation rule can be found here: https://en.wikipedia.org/wiki/International_Securities_Identification_Number

### PR DESCRIPTION
**Pull request by VincentLanglet**
Co-Authored-By: Yannis Foucher <33806646+YaFou@users.noreply.github.com>

| Q             | A
| ------------- | ---
| Branch?       | master
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | no
| Tickets       | Fix #36362
| License       | MIT
| Doc PR        | symfony/symfony-docs#13960

Rebase of https://github.com/symfony/symfony/pull/36368

I asked him by mail and he didn't have time to finish the PR and allowed me to do it.

### CODE REVIEW

File: src/Symfony/Component/Validator/Tests/Constraints/IsinValidatorTest.php [hunk 1]
```diff
@@ -0,0 +1,221 @@
...
+    public function testValidIsin($isin)
+    {
+        $this->validator->validate($isin, new Isin());
+        $this->assertNoViolation();
+    }
+
+    public function getValidIsin()
+    {
+        return [
+            ['XS2125535901'], 	// Goldman Sachs International	MTN	HKD HKD	126d
```
Comment (javiereguiluz) [inline src/Symfony/Component/Validator/Tests/Constraints/IsinValidatorTest.php:42]: These text fixtures doesn't look very useful ... they are all `X....`  What if we remove most of them and add instead some of the examples of the Isin Wikipedia page?

```php
['US0378331005'],   // Apple, Inc.
['AU0000XVGZA3'],   // TREASURY CORP VICTORIA 5 3/4% 2005-2016
['GB0002634946'],   // BAE Systems
```

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php [hunk 2]
```diff
@@ -0,0 +1,129 @@
...
+        if (!preg_match(Isin::VALIDATION_PATTERN, $value)) {
+            $this->context->buildViolation($constraint->message)
+                ->setParameter('{{ value }}', $this->formatValue($value))
+                ->setCode(Isin::INVALID_PATTERN_ERROR)
+                ->addViolation();
+
+            return;
+        }
+
+        if (!$this->isCorrectChecksum($value)) {
```
Comment (javiereguiluz) [inline src/Symfony/Component/Validator/Constraints/IsinValidator.php:63]: At this point in code, `$value` is a `string` so we may add some type-hints and return types.

**Review by javiereguiluz — APPROVED** [review]

File: src/Symfony/Component/Validator/Tests/Constraints/IsinValidatorTest.php (hunk 1 above)
Comment (VincentLanglet) [reply src/Symfony/Component/Validator/Tests/Constraints/IsinValidatorTest.php:42]: Done.

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php (hunk 2 above)
Comment (VincentLanglet) [reply src/Symfony/Component/Validator/Constraints/IsinValidator.php:63]: Done.

File: src/Symfony/Component/Validator/CHANGELOG.md [hunk 3]
```diff
@@ -28,6 +28,7 @@ CHANGELOG
     * })
     */
    ```
+ * added `Isin` constraints
```
Comment (derrabus) [inline src/Symfony/Component/Validator/CHANGELOG.md:31]:
```suggestion
 * added the `Isin` constraint and validator
```

File: src/Symfony/Component/Validator/Tests/Constraints/IsinValidatorTest.php [hunk 4]
```diff
@@ -0,0 +1,132 @@
...
+            ['DE000HZ8VA77'], 	// UniCredit Bank AG	Bond	EUR EUR	94d
+            ['CH0528261156'], 	// Leonteq Securities AG [Guernsey]	Bond	GBP GBP	3y
+            ['US0378331005'],   // Apple, Inc.
+            ['AU0000XVGZA3'],   // TREASURY CORP VICTORIA 5 3/4% 2005-2016
+            ['GB0002634946'],   // BAE Systems
+            ['CH0528261099'], 	// Leonteq Securities AG [Guernsey]	Bond	EUR EUR	3y
+            ['XS2155672814'], 	// OP Corporate Bank plc	CP	GBP GBP	93d
+            ['XS2155687259'], 	// Orbian Financial Services III, LLC	CP	USD USD	206d
+            ['XS2155696672'], 	// Sheffield Receivables Company LLC	CP	EUR EUR	7d//Vatican City State
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Tests/Constraints/IsinValidatorTest.php:51]: please use only spaces, no tabs

File: src/Symfony/Component/Validator/Resources/translations/validators.fr.xlf [hunk 5]
```diff
@@ -382,6 +382,10 @@
                 <source>Each element of this collection should satisfy its own set of constraints.</source>
                 <target>Chaque élément de cette collection doit satisfaire à son propre jeu de contraintes.</target>
             </trans-unit>
+            <trans-unit id="99">
+                <source>This value is not a valid International Securities Identification Number (ISIN).</source>
+                <target>Ce n'est pas un code international de sécurité valide (ISIN).</target>
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Resources/translations/validators.fr.xlf:387]: Cette valeur ...

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php [hunk 6]
```diff
@@ -0,0 +1,125 @@
...
+        // get the total value of all the digits
+        $sum = array_sum($numbers);
+
+        // get the remainder when dividing by 10
+        $mod = $sum % 10;
+
+        // subtract from 10
+        $rem = 10 - $mod;
+
+        // mod from 10 to catch if the result was 0
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Validator/Constraints/IsinValidator.php:120]: I'm not sure these comments above add value, the code tells the same

Comment (VincentLanglet) [reply src/Symfony/Component/Validator/Constraints/IsinValidator.php]: I remove some comments. WDYT now ?

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php [hunk 7]
```diff
@@ -0,0 +1,118 @@
...
+        $number = implode('', $characters);
+        $expectedCheckDigit = $this->getCheckDigit($number);
+
+        return $checkDigit === $expectedCheckDigit;
+    }
+
+    /**
+     * This method performs the Luhn algorithm to obtain a check digit.
+     */
+    private function getCheckDigit(string $input): int
```
Comment (fabpot) [inline src/Symfony/Component/Validator/Constraints/IsinValidator.php:87]: Instead, we should probably use the Luhn validator, that would avoid having 2 implementations of the same algo in code.

**Review by fabpot — CHANGES_REQUESTED** [review]

File: src/Symfony/Component/Validator/Constraints/IsinValidator.php (hunk 7 above)
Comment (VincentLanglet) [reply src/Symfony/Component/Validator/Constraints/IsinValidator.php]: Indeed. I made the changes. Is it ok the way I did @fabpot ?

**Review by fabpot — APPROVED** [review]

**Comment by fabpot** [comment]
Thank you @VincentLanglet.

**Pull request merged**